### ISSUE TITLE
CORS headers missing

### ISSUE THREAD

**Issue by greggman** [soundcloud/api#100]
## Title: Cannot call /resolve
Issue found of: August 7, 2021

### Endpoint(s):
* `GET /resolve`

### Scope(s):
* None (application is not using authentication i.e., implicit flow)

### Steps to reproduce:
```bash
curl --verbose  https://api.soundcloud.com/resolve\?url\=https%3A%2F%2Fsoundcloud.com%2Fchibi-tech%2Flolitazia-season\&client_id\=3f4914e324f9caeb23c521f0f1835a60\&format\=json\&_status_code_map%5B302%5D\=200
```

### Expected behaviour:
CORS headers are returned in the headers so the browser allows JavaScript to read the response

```
Access-Control-Allow-Origin: *
```

### Actual behaviour:
No CORS headers are returned

**Comment by murphinator** [comment]
Having the same issue since last week.

**Comment by jasongrishkoff** [comment]
Hey guys, this affected me last week as well. The solution is to use the new auth headers when requesting your endpoint.  `client_id` is no longer a valid URL parameter.
More details here: https://developers.soundcloud.com/blog/security-updates-api

**Comment by greggman** [comment]
If I understand correctly, auth headers require the user to login but my site doesn't require the user to login so that's not a solution

**Comment by jasongrishkoff** [comment]
Nope, for the resolve endpoint it's *your* client ID and client secret that get used to create the auth token.  User doesn't have to be connected unless you're trying to access their personal endpoints.  Resolving a URL or getting a track's API data don't require the user to authenticate -- they require *you* to authenticate as the developer.

**Comment by greggman** [comment]
Well, that's slightly better. The issue is my app was doing all soundcloud stuff client side only.

So it sounds like what I have to do now is add a bunch of new endpoints on my server so that the webpage can ask the server for an auth token which the webpage can then use to make the calls it was making before. Further the pages have to ping my server all the time because the auth tokens expire in 6 hours.

Why? No other service I know requires auth for public data. Example, here's me asking github for a list of soundcloud's public repos

```
curl \
  --verbose -H "Accept: application/vnd.github.v3+json" \
  https://api.github.com/users/soundcloud/repos
```

It provides the data and sends the correct cors headers.

the data I'm asking for is publically available by scrapping the site so gating it on auth seems kind of silly.

**Comment by dasha-kobzeva** [comment]
Hi @greggman,
thank you for filing the issue.

Are you providing an `Origin` header in your request? In order for CORS headers to work, the `Origin` header must be present. Could you please provide it in the request and let us know if it resolves the issue?

Regarding the `Auth` header, as was correctly mentioned you can use the `client_credentials` flow which does not require the user, only your application. `Client_id` is replaced with a header in this case.

We are tightening the security due to the impersonation.
With the current implementation, we cannot verify that requests coming from your app actually come from it, or someone is abusing your `client_id`.

**Comment by svobik7** [comment]
Why @dasha-kobzeva there is no client side way of getting the token? Like @greggman describes, this change is forcing us to add some server side stuff which were not required before in pretty short period - email about deprecating client_id on 31th July I received on 2nd August 👎

**Comment by dasha-kobzeva** [comment]
Dear @svobik7,
client-side authentication is not secure. 
We cannot trust the client as an attacker could have modified the client as they wish. Our goal is to make the API more secure and ensure the clients that using it are the ones that our services trust.

We announce all breaking changes in our [blog](https://developers.soundcloud.com/blog/), which you are welcome to subscribe to. We also make announcements in our [Twitter](https://twitter.com/soundclouddev) account.

**Comment by JackBrit** [comment]
I've successfully managed to use the `client_credentials` flow to get an authentication token and used that to then continue using the /resolve endpoint. However this only appears to work in Chrome. When I attempted to load the same app in Safari i get an `XMLHttpRequest cannot load https://api.soundcloud.com/tracks/XXXXXXXX due to access control checks.`

This appears to be an issue with the headers being returned by the Soundcloud server? Any idea on a fix?

**Comment by dasha-kobzeva** [comment]
> I've successfully managed to use the `client_credentials` flow to get an authentication token and used that to then continue using the /resolve endpoint. However this only appears to work in Chrome. When I attempted to load the same app in Safari i get an `XMLHttpRequest cannot load https://api.soundcloud.com/tracks/XXXXXXXX due to access control checks.`
> 
> This appears to be an issue with the headers being returned by the Soundcloud server? Any idea on a fix?

Hi @JackBrit,
we rolled out the fix, could you verify if you are still seeing the issue?

**Comment by JackBrit** [comment]
@dasha-kobzeva I'm still seeing the issue in Safari. The app is using SDK version `3.3.2`

**Comment by dasha-kobzeva** [comment]
> @dasha-kobzeva I'm still seeing the issue in Safari. The app is using SDK version `3.3.2` and the Safari version I'm using is `14.1.1`

the issue seems to persist only in Safari. We need some time to investigate.

**Comment by greggman** [comment]
I am not seeing the correct CORS headers even with auth

```
curl --request POST \
--url https://api.soundcloud.com/oauth2/token \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data client_id=3f4914e324f9caeb23c521f0f1835a60 \
--data client_secret=<secret-was-here> \
--data grant_type=client_credentials
```

returns 

```
{"access_token":"2-157079--BeMBan<token-cont>","expires_in":3599,"refresh_token":"EcDGIj2AWazvIAjkD<token-cont>","scope":""}
```

ok so we use the token

```
curl --verbose  https://api.soundcloud.com/resolve\?url\=https%3A%2F%2Fsoundcloud.com%2Fchibi-tech%2Flolitazia-season\&format\=json\&_status_code_map%5B302%5D\=200 --header 'Authorization: OAuth 2-157079--BeMBan<token-cont>'
```

and the response is

```
< HTTP/1.1 302 Moved Temporarily
< Content-Type: application/json; charset=utf-8
< Content-Length: 82
< Connection: keep-alive
< Date: Tue, 10 Aug 2021 14:48:52 GMT
< Via: 1.1 67c5370c4ea2e5f9f865960e31ef2d86.cloudfront.net (CloudFront), 1.1 67c5370c4ea2e5f9f865960e31ef2d86.cloudfront.net (CloudFront)
< Host: api.soundcloud.com
< Accept: application/json
< Location: https://api.soundcloud.com/tracks/113332220
< X-Real-IP: 98.207.186.238
< User-Agent: curl/7.64.1
< x-client-id: 
< x-request-id: 611291D4:128B:0:1E20498
< x-robots-tag: noindex
< Authorization: OAuth 2-157079--BeMBan<token-cont>
< Cache-Control: private, max-age=0
< x-ssl-details: TLSv1.2/ECDHE-RSA-AES128-GCM-SHA256
< x-consumer-key: 
< referrer-policy: no-referrer
< x-ampelmann-cdn: cf
< X-Forwarded-For: 98.207.186.238
< x-frame-options: DENY
< x-request-start: t=1628606932.000
< X-Forwarded-Proto: https
< x-content-type-options: nosniff
< x-client-credentials-id: 157079
< cloudfront-viewer-country: US
< cloudfront-forwarded-proto: https
< strict-transport-security: max-age=63072000; includeSubdomains; preload
< Server: am/2
< Vary: Origin
< X-Cache: Miss from cloudfront
< X-Amz-Cf-Pop: SFO20-C1
< X-Amz-Cf-Id: ZNjwprk8OftP1-gT67MZ6FWCd8xzXHlieTRrGoz_FWTVWCVtGCan9g==
< 
* Connection #0 to host api.soundcloud.com left intact
{"status":"302 - Found", "location":"https://api.soundcloud.com/tracks/113332220"}* Closing connection 0

```

There is no `Access-Control-Allow-Origin: *` header in there. 

Is it just impossible now to use soundcloud's API from the browser period?

**Comment by greggman** [comment]
I would like to make sure I understand how to solve this issue. My current thought is

1. client (the browser), asks server (my site) for an auth token
2. server (my site), if it has a non-expired token returns it, else it makes the client_credientials call to soundcloud and gets an auth token, passes that back to client. It can pass that same token many clients.
3. clients can now make calls until the auth token expires
4. for the client, when token expires go to step 1

Is that a viable solution?

What's not clear is when I need to do the refresh call. That part makes no sense to me unless I pass the old auth token back (which according to the docs I don't). So what is the point of the refresh call vs just calling client_credientials again?

**Comment by JackBrit** [comment]
> I would like to make sure I understand how to solve this issue. My current thought is
> 
> 1. client (the browser), asks server (my site) for an auth token
> 2. server (my site), if it has a non-expired token returns it, else it makes the client_credientials call to soundcloud and gets an auth token, passes that back to client. It can pass that same token many clients.
> 3. clients can now make calls until the auth token expires
> 4. for the client, when token expires go to step 1
> 
> Is that a viable solution?
> 
> What's not clear is when I need to do the refresh call. That part makes no sense to me unless I pass the old auth token back (which according to the docs I don't). So what is the point of the refresh call vs just calling client_credientials again?

@greggman That was essentially my approach. Although I found that I still needed to include my client ID as a param with the stream and download URLs following a successful /resolve request - Not sure if these are part of your app.

**Comment by JackBrit** [comment]
> > @dasha-kobzeva I'm still seeing the issue in Safari. The app is using SDK version `3.3.2` and the Safari version I'm using is `14.1.1`
> 
> the issue seems to persist only in Safari. We need some time to investigate.

Thanks for investigating @dasha-kobzeva. Not sure if it will aid you/your team but [here](https://controlc.com/d8ba4929) are the response headers I get from a successful request (Auth token redacted).

I don't get any headers response headers within Safari.

**Comment by dasha-kobzeva** [comment]
Hi @JackBrit,
thank you for the detailed info. i can also reproduce the issue.

Could you please tell me if it ever worked for you in Safari? Because it seems to me even the old version had the same issue in Safari.

Thank you

**Comment by JackBrit** [comment]
> Hi @JackBrit,
> thank you for the detailed info. i can also reproduce the issue.
> 
> Could you please tell me if it ever worked for you in Safari? Because it seems to me even the old version had the same issue in Safari.
> 
> Thank you

Hi @dasha-kobzeva 
Yes this was previously working in Safari. I was originally using SDK version 3.1.2. if that helps at all.

**Comment by greggman** [comment]
> As far as I understand a token created via client credentials have a 6 hour expiry time. Therefore you would need to timestamp the token and check if it has expired and then call the refresh token endpoint.

I don't understand this. refresh_token returns a new auth token. So what's the point of calling refresh vs calling client_credentials and getting a new auth token. Both paths given a new auth token

**Comment by mgoodfellow** [comment]
> I don't understand this. refresh_token returns a new auth token. So what's the point of calling refresh vs calling client_credentials and getting a new auth token. Both paths return a new auth token

AFAIK in this particular use case, it doesn't matter which you choose - obviously in the case of a user login flow via /connect, you want to use the refresh to avoid them having to login every 6 hours - but with the client_credentials flow, it's just the server authenticating for public endpoints. You can just do that process on demand as required instead of storing and maintaining a refresh token.

**Comment by dasha-kobzeva** [comment]
> > Hi @JackBrit,
> > thank you for the detailed info. i can also reproduce the issue.
> > Could you please tell me if it ever worked for you in Safari? Because it seems to me even the old version had the same issue in Safari.
> > Thank you
> 
> Hi @dasha-kobzeva
> Yes this was previously working in Safari. I was originally using SDK version 3.1.2. if that helps at all.

Hi, which SDK you are talking about? Soundcloud JS SDK?

**Comment by hanoii** [comment]
I am also having CORS issues:

<img width="1011" alt="Screen Shot 2021-08-12 at 10 39 28" src="https://user-images.githubusercontent.com/677879/129206821-54627a80-392c-4d89-8348-bd5050273d59.png">

It seems like the first resolve URL is working (I can see CORS headers) but then the redirect fails as the new URL doesn't have the CORS headers.

Is this also due to the recent upgrade.

And I don't understand something, there's no more pure client-side integrations that we can use. We'll always have to post it to the server?

Also, what happens if the client is the one to get the refresh token and we make the client secret available on the JS side of things. I know that might sound as a bad practice but if it's still possible then we would be at the same situation as before. Why not make public resources API endpoints free to use? Or at least enable them for specific referrers. That's how most API works (google maps for instance).

**Comment by dasha-kobzeva** [comment]
Hi @hanoii,

you are getting an error on /resolve redirect because you are not authenticated properly.

You send a request to /resolve, providing client_id. When the resolution happens, you get a 302 response with the next URL to call, for instance, /users/some_id. That second call was stripped down from client_id, so our API treats it as unauthenticated.

To avoid such behavior your call to /resolve must contain an Authorization header so that the redirect can also pick it up and give you a proper response back.

Answering your question `Why not make public resources API endpoints free to use?` we want to make our API more secure and be sure that the request coming from your app is actually coming from it. Exposing API keys might lead to leakage and impersonated traffic. 
We want to protect our clients and also our platform from any abuse. 
Yes, with new updates pure client-side integration is not as possible as it used to be. We are considering updating our SDK for those exact purposes, but I cannot give you any defined timelines/plans at the moment.

Hope that answers your question.

**Comment by hanoii** [comment]
@dasha-kobzeva.

In my case I am only using resolve to get some track info as well as the streaming url and then playing the stream through an angular module that receives the URL. I am also sending `?client_id` on the stream URL, would that also fail to work.

So if I understand correctly, now to to use the resolve endpoint I would need to implement a server side endpoint on my server that fetches a token which then the client can use to use the resolve endpoint? Is that so?

How would I have to work around for the streaming url to work?

And by free, i mean no api key. They are, after all, public resources. What difference there is in terms of abuse between querying it through the API or accessing www.soundcloud.com?

**Comment by dasha-kobzeva** [comment]
> 
>  I am also sending `?client_id` on the stream URL, would that also fail to work?
 By the end of September, using `?client_id` won't be sufficient anymore. Every request to `api.soundcloud.com`, streaming included, requires an Auth header now. 

Yes, to comply with the recent changes and be able to integrate with SC you have to update your application.

>
>And by free, i mean no api key. They are, after all, public resources. What difference there is in terms of abuse between querying it through the API or accessing www.soundcloud.com?
>

I believe this is a bigger topic for discussion, so I would like to skip it at this point.

**Comment by greggman** [comment]
> streaming included, requires an Auth header now.

How do I include a header in an HTML `<audio>` tag for streaming?

**Comment by jasongrishkoff** [comment]
@greggman you can't. See #93.

**Comment by JackBrit** [comment]
> > > Hi @JackBrit,
> > > thank you for the detailed info. i can also reproduce the issue.
> > > Could you please tell me if it ever worked for you in Safari? Because it seems to me even the old version had the same issue in Safari.
> > > Thank you
> > 
> > 
> > Hi @dasha-kobzeva
> > Yes this was previously working in Safari. I was originally using SDK version 3.1.2. if that helps at all.
> 
> Hi, which SDK you are talking about? Soundcloud JS SDK?

Hi @dasha-kobzeva sorry, missed this amongst the other comments. Yes, I'm referring to the JS SDK

**Comment by dasha-kobzeva** [comment]
@JackBrit 
I'm afraid the JS SDK is not gonna work for this endpoint at the moment. I would suggest calling API directly

**Comment by JackBrit** [comment]
> @JackBrit
> I'm afraid the JS SDK is not gonna work for this endpoint at the moment. I would suggest calling API directly
@dasha-kobzeva 
Ah sorry I am calling the /resolve endpoint directly and not using the SDK - That was legacy code.
I've also updated my codebase to use authentication to obtain the streamable MP3s. However, this is hasn't resolved the CORS issue within Safari.

**Comment by greggman** [comment]
I'm trying to understand the `"expires_in":3599` field. The blog post linked above says TTL is 6 hours. Not sure how to convert 3599 into 6 hours. Is the blog post wrong and the expire in 1 hour?

**Comment by svobik7** [comment]
> > @JackBrit
> > I'm afraid the JS SDK is not gonna work for this endpoint at the moment. I would suggest calling API directly
> 
> @dasha-kobzeva
> Ah sorry I am calling the /resolve endpoint directly and not using the SDK - That was legacy code.
> I've also updated my codebase to use authentication to obtain the streamable MP3s. However, this is hasn't resolved the CORS issue within Safari.

Same for me. I managed to make our integration working again in Chrome with no problems but in Safari. Here is what I am getting instead of track data:
<img width="1423" alt="Screenshot 2021-08-16 at 9 22 20" src="https://user-images.githubusercontent.com/761766/129526381-c61e73e3-31b5-4185-b7f0-9aeeb19a7cb3.png">
<img width="1008" alt="Screenshot 2021-08-16 at 9 22 34" src="https://user-images.githubusercontent.com/761766/129526385-c51170fc-1e37-4947-bebf-8943edef7ccc.png">

**Comment by rahul-sc** [comment]
> I'm trying to understand the `"expires_in":3599` field. The blog post linked above says TTL is 6 hours. Not sure how to convert 3599 into 6 hours. Is the blog post wrong and the expire in 1 hour?

That's true - good catch. The blog is wrong. TTL for tokens is 1 hour

**Comment by dasha-kobzeva** [comment]
@svobik7 @JackBrit,
the issue with Safari has existed for a long time now, it's not newly introduced. JS SDK had a workaround specifically for it, maybe it can be of use to you as we have decided to stop maintaining SDK in the future. 

```
      /*
       * Tell the API not to serve a redirect. This is to get around
       * CORS issues on Safari 7+, which likes to send pre-flight requests
       * before following redirects, which has problems.
       */
      _status_code_map: { 302: 200 }
```
So the idea is to overwrite the response status from 302 to 200 so you get redirect back, instead of the resource.
I hope that helps.

**Comment by svobik7** [comment]
@dasha-kobzeva , no luck for me :-( I tried to add `_status_code_map` to my api call in the same way SDK does but still getting CORS error in safari. Were you able to manage it @JackBrit ?

**Comment by dasha-kobzeva** [comment]
> @dasha-kobzeva , no luck for me :-( I tried to add `_status_code_map` to my api call in the same way SDK does but still getting CORS error in safari. Were you able to manage it @JackBrit ?

Do you mean you added it as a query parameter? I believe that won't work as API does not know about it. 
Maybe the convo in this thread might be of use: https://github.com/soundcloud/api/issues/93
You should intercept the 302 request and treat it as 200 so Safari does not try to resolve it.

**Comment by svobik7** [comment]
@dasha-kobzeva I dont think this possible in browser as there is no way how to get the 302 location in current fetch libs. It would be great if resolve endpoint just return 200 with additional info. Is that possible?

**Comment by greggman** [comment]
https://stackoverflow.com/a/57239126

**Comment by aproni34f** [comment]
> Answering your question `Why not make public resources API endpoints free to use?` we want to make our API more secure and be sure that the request coming from your app is actually coming from it. Exposing API keys might lead to leakage and impersonated traffic.
> We want to protect our clients and also our platform from any abuse.
> Yes, with new updates pure client-side integration is not as possible as it used to be. We are considering updating our SDK for those exact purposes, but I cannot give you any defined timelines/plans at the moment.
> 
> Hope that answers your question.

Youtube provides api keys and they offer an option to lock a key to a domain. Why dont you do the same?

**Comment by steam0r** [comment]
hey,

not sure if this is releated, but we over at cables.gl are trying to move over to the new soundcloud api to continue giving users the possibility of easy audio access in their cables-patches. our basic workflow is like this:

- ask user for clientid, secret and any soundcloud url
- get token: https://api.soundcloud.com/oauth2/token
- try to resolve track at the given url: https://api.soundcloud.com/resolve.json?url=<url>
- request track information (actually being redirected here by the above request): https://api.soundcloud.com/tracks/<trackId>
- request track stream: https://api.soundcloud.com/tracks/<trackId>/stream

all of these requests work fine, except for the last one that gives "CORS header 'Access-Control-Allow-Origin' missing". and a 401 unauthorized header.

is this workflow even possible any longer with the new api, am i doing something wrong? browser is firefox (on mac) and
response-headers look like this:

![Bildschirmfoto 2021-08-19 um 16 38 06](https://user-images.githubusercontent.com/151271/130088249-fb366dd9-6f45-4dbd-9257-0d0dfe0509b8.png)

with my request being:

![Bildschirmfoto 2021-08-19 um 16 38 40](https://user-images.githubusercontent.com/151271/130088374-d28d2926-b2f7-49de-9c56-c15d2eacc27b.png)

any help would be appreciated. thanks!

**Comment by JackBrit** [comment]
> * request track stream: https://api.soundcloud.com/tracks/$trackId/stream

@steam0r I believe this URL should be https://api.soundcloud.com/tracks/$trackId/streams (note the added "s" to stream)

**Comment by steam0r** [comment]
thanks for your reply, i am not manually constructing this url though, it's `stream_url` thats provided by https://api.soundcloud.com/tracks/$trackId

**Comment by greggman** [comment]
I would be really helpful if Soundcloud would post a working example!!!

In particular this used to work (using client_id=...)

https://jsgist.org/?src=9a66a5082ce23a5f77d9c961b3d9b3fe

Now though I ask for a token, then I include it in the header for `/resolve`. This gives me a `stream_url` which I then use fetch `method: HEAD`, again including the token, to get a redirect URL but it just gives me the same URL. Checking the response headers there is no `location` in `access-control-allow-headers` so that I can read the Location header so I can't get the URL.

```
HTTP/1.1 302 Moved Temporarily
Content-Type: text/plain; charset=utf-8
Connection: keep-alive
Date: Sat, 21 Aug 2021 09:58:16 GMT
Location: https://cf-media.sndcdn.com/Rh54NAB0GOIj.128.mp3?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiKjovL2NmLW1lZGlhLnNuZGNkbi5jb20vUmg1NE5BQjBHT0lqLjEyOC5tcDMqIiwiQ29uZGl0aW9uIjp7IkRhdGVMZXNzVGhhbiI6eyJBV1M6RXBvY2hUaW1lIjoxNjI5NTQwMzU0fX19XX0_&Signature=ALkdUZWGZ5vAq2NzbvV67phK5W5X6ZcwLbnoOqrPYpSji8mdmTFaxfEbx7z0gkju0PYf5nSpqQRjRY-qnLcm5pzH5N4Hhw5sCxphCx9vyFF5x8moG7nYjtPa-PKdGYMubV1A4lM0-0v1COXOb6KS3nYHMUpita33KqsD8p2jpwpw3s-wgcXbmDU7y8OqIY4F9uP7woaY1LdQlgZyiR8Q2RURRnT2pbK8zip8moGHosHchD06m~MXHOZOiBNsZa7TAayTqtYpC-6LnqOKsoQ~lCw0mzv2IzjF0dwLbwIhQZzchxCxhqGQr3NhtpcXjE6ovkfolEF-Lxrg9Nm7gZrKzw__&Key-Pair-Id=APKAI6TU7MMXM5DG6EPQ
x-robots-tag: noindex
referrer-policy: no-referrer
x-frame-options: DENY
access-control-max-age: 1728000
x-content-type-options: nosniff
access-control-allow-origin: http://localhost:3000
access-control-allow-headers: Authorization, Content-Type, Device-Locale, X-CSRF-Token
access-control-allow-methods: DELETE, GET, PATCH, POST, PUT
access-control-expose-headers: Date
access-control-allow-credentials: true
strict-transport-security: max-age=63072000; includeSubdomains; preload
Server: am/2
Vary: Origin
X-Cache: Miss from cloudfront
Via: 1.1 a3b511649bade26170091701ae26b616.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SFO53-C1
X-Amz-Cf-Id: CIfiUH7liqMQQ6rkIMoc_o2yUUz_hmDT8srkuBID9c446pOjAHOVJw==
```

If I pass the original stream URL to an audio tag I get

> Access to audio at 'https://api.soundcloud.com/tracks/63778891/stream' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

**Comment by greggman** [comment]
In particular the site I'm trying to fix to use the new API does audio reactive visualizations like this

https://jsfiddle.net/greggman/01vj5y6L/

Is there any chance you could show a working example of doing the same using the new API?

**Comment by greggman** [comment]
Here's a minimal node/express based server and audio-reactive example that does NOT work. Could someone from Soundcloud please take a look and tell us the correct way to make this work?

https://github.com/greggman/soundcloud-audio-reactive-example

**Comment by steam0r** [comment]
+1 for "a working example", what @greggman describes is exactly what we try and fail to accomplish with the new API

**Comment by sanjutagra** [comment]
Facing the same issue, it would be great if Soundcloud team can provide a working example or look at the example @greggman shared above and suggest the fix.

**Comment by mgoodfellow** [comment]
I had a few mins so I PR'ed a working fix into @greggman 's example - apologies it's a tad hacky - https://github.com/greggman/soundcloud-audio-reactive-example/pull/1

Anyway, the trick is using the documented form of /tracks/{id}/streams here: https://developers.soundcloud.com/docs/api/explorer/open-api#/tracks/get_tracks__track_id__streams instead of using the /tracks/{id}/stream which is returned on the SoundCloud track object.

This method means you aren't using a redirect to get the CF stream URL which seems to be causing the CORS issues with CloudFront.

**Comment by steam0r** [comment]
thanks for that @mgoodfellow, works for us. this is somewhat different than what we did up until now and also relies on the urls of the API not changing. i think we'll go with this now, it would be reassuring if soundcloud could confirm this is a good way of doing this, though.

**Comment by mgoodfellow** [comment]
That is the documented API endpoint for retrieving stream endpoints from the API Explorer, so I would be confident that you can rely on it not changing. This is the API contract that SoundCloud provides to us. 

Interestingly, the /stream endpoint is not documented, but I am aware that that format is returned on the SoundCloud track metadata object. It would be good to also confirm why the /stream endpoint is not present in the API documentation, and if there are plans for that endpoint to be deprecated in future.

**Comment by greggman** [comment]
Thank you! I will try it.

As for the /stream endpoint, IIRC that flow was what the SoundCloud 3.0
library did

**Comment by dasha-kobzeva** [comment]
Hi,
it seems as the issue is resolved. As the thread grew up quite big I will close it. In case you are facing any issues, please create a separate ticket.

Thank you

**Comment by JackBrit** [comment]
@dasha-kobzeva I don't think this issue should have been closed or moved to done on your Kanban board as if you follow @mgoodfellow's generous investigation his solution still does not work on Safari.

This really needs to be resolved on Soundclouds API side of things and with the amount of money spent on Soundcloud accounts its a bit of a joke to rely on users to fix Soundcloud issues with hacky workarounds.

Could you please reopen this ticket and begin to work on a fix. Thank you

**Comment by mgoodfellow** [comment]
Well, @gregmann did all the work on building a reproducible sample :)

In this particular case, I would say that it feels like more of an issue with Safari seeing as it is the odd one out. Not to put words into SoundCloud's mouth, but my understanding is that they are very much advising against pure client-side implementations. In fact, a true pure client-side solution is no longer possible as it would leak the `client_secret`. This means you will always have to have a server-side implementation going forward.

If you look at my discussion with @gregmann on my [PR against his repo](https://github.com/greggman/soundcloud-audio-reactive-example/pull/1), I actually recommend pushing all the API calls to the server side (which is what I do with my own implementations) and avoid calling the SoundCloud API from your client-side implementation. This would solve the issue with Safari, and would still only consist of the same components.

Allowing the token to leak to the frontend doesn't feel "right" or secure from an implementation point of view, and since you already have the server-side implementation to generate the token, you may as well proxy the full set of calls to happen on your server behind your own convenience route (in that example I give `/resolve-stream-url?url=XXX` as an example), and simply return the final bit of data you wanted to your client (in this case, the signed stream URL for CloudFront). This avoids all the pain of redirects with CORS and specific browser implementations behaving differently with the token being passed in the header.

**Comment by greggman** [comment]
Just to follow up, I updated [the example](https://github.com/greggman/soundcloud-audio-reactive-example) to do everything on the server. It should work in Safari.

In my actual project I have 2 endpoints, one `/resolve?url=xxx` which the server makes the call to soundcloud and returns the results back to the page. another is `/track_url` which given a stream url returns the url to pass to the `<audio>` tag.  The reason it's 2 endpoints instead of 1 is `/resolve` can return info for a tracklist and the page needed that info.

The code is all hacked and I'm embarrassed to show it but [it's working](https://github.com/greggman/vertexshaderart) on Safari both desktop and iOS at the moment.

**Comment by salvatorespoto** [comment]
It seems to happen to me right now, using Vivaldi and Chrome browsers. The request to the SoundCloud API gets "401 - Unauthorized"
